File: Semester4/Advanced Programming/Lab2_Hw/src/main/java/repository/CRUDRepository.java
package repository;

import domain.IHasId;
import domain.exceptions.IllegalArgumentException;
import domain.exceptions.ValidatorException;

public interface CRUDRepository<ID, E extends IHasId<ID> > {
    /**
     * Function to find an element into the repository
     * @param id - id of the entity to be remove
     * @return - returns the entity with the given id, or
     * null if there is no entity with the specified id
     * @throws - Illegal argument exception if
     */
    E findOne(ID id) throws IllegalArgumentException;

    /**
     *
     * @param entity - entity to be saved
     * @return - Null if the entity was added, the entity otherwise
     * @throws IllegalArgumentException - if entity is null
     * @throws ValidatorException - if the entity is not well defined
     */
    E save(E entity) throws IllegalArgumentException, ValidatorException;

    /**
     * Function to delete an entity from the repository
     * @param id - id of the entity
     * @return - the entity if it was deleted, null otherwise
     * @throws IllegalArgumentException - if the id is null
     */
    E delete(ID id) throws IllegalArgumentException;

    /**
     * Function to update an entity from the repository
     * @param entity - the entity to be updated
     * @return - Null if the entity was updated, the given entity otherwise
     * @throws IllegalArgumentException - if the entity is null
     * @throws ValidatorException - if the entity is not well defined
     */
    E update(E entity)throws IllegalArgumentException, ValidatorException;

    /**
     * Function to get all the entities from the repository
     * @return - an iterable containing all the elements from the repository
     */
    Iterable<E> findAll();

    /**
     * Function which returns the size of the repository
     * @return - integer representing number of elements from the repository
     */
    Integer size();
}

File: Semester3/Advanced Programming Methods/Labs/StudentsManagementAppC#/StudentsManagementApp/repository/CrudRepository.cs
using System.Collections.Generic;
using System.Linq;

namespace StudentsManagementApp.repository
{
	/// <summary>
	/// CRUD operations repository interface
	/// </summary>
	/// <typeparam name="TId"> TE must have an attribute of type TId</typeparam>
	/// <typeparam name="TE">Type of elements saved in Repository</typeparam>
	public interface ICrudRepository<TId, TE>
	{
		/// <summary>
		/// Function to find an element into the repository
		/// </summary>
		/// <param name="id">id of the entity to be returned</param>
		/// <returns>The element if it exists, null otherwise</returns>
		TE FindOne(TId id);

		/// <summary>
		/// Function to get all elements of the repository as Enumerable
		/// </summary>
		/// <returns>Enumerable containing all entities from the repository</returns>
		IEnumerable<TE> FindAll();

		/// <summary>
		/// Function to save an entity to the repository
		/// </summary>
		/// <param name="entity">entity to be saved</param>
		/// <returns>null if the entity was saved, the given entity otherwise</returns>
		TE Save(TE entity);


		/// <summary>
		/// Function to delete an entity from the repository
		/// </summary>
		/// <param name="id">id of the entity to be deleted</param>
		/// <returns>the deleted entity if it could be deleted, null otherwise</returns>
		TE Delete(TId id);

		/// <summary>
		/// Function to update an entity from the repository
		/// </summary>
		/// <param name="entity">new entity</param>
		/// <returns>null if the entity was updated, the given entity otherwise</returns>
		TE Update(TE entity);
	}
}